### ISSUE TITLE
Sharing challenge via QR code or NFC

### ISSUE THREAD

**Issue by frief** [lexica/lexica#102]
it would be great if a challenge could be shared with (one or more) others.

Both QR code and NFC have the advantage that the players would need to directly interact.
(Generates extra dynamics that WLAN cannot not give)

Appending QR-codes for a 4x4 and a 6x6 quiz.

Again, thanks for Lexica!

"lexica:ESSNNRVOOREITGAA"
![4x4](https://user-images.githubusercontent.com/2320349/48660126-e4b8b080-ea5c-11e8-8e57-5f30f1768c1f.png)

"lexica:SHAUNUOJIASHIANAHPHATIOSHAOSOHYEICHE"
![6x6](https://user-images.githubusercontent.com/2320349/48660127-e4b8b080-ea5c-11e8-9467-7af6f7660c75.png)

**Comment by d-mal** [comment]
Great idea!

**Comment by pserwylo** [comment]
As mentioned in #213, this is of course still a good idea, but it needs to take into account the type of game being played, not just the board. It is worth noting that there is already a mechanism for:
* Saving game modes to strings for serialization ([`GameSaverPersistent`](https://github.com/lexica/lexica/blob/63bdf4532dc57463642c47d3bb2d7cedc30467df/app/src/main/java/com/serwylo/lexica/GameSaverPersistent.java#L81) and [`GameMode.serialize()`](https://github.com/lexica/lexica/blob/63bdf4532dc57463642c47d3bb2d7cedc30467df/app/src/main/java/com/serwylo/lexica/db/GameMode.kt#L90)). Probably needs work to make it more compact for sharing though.
* Creating a new game mode when we come across a series of game mode options that don't yet exist ([`findMatchingGameModeId()`](https://github.com/lexica/lexica/blob/master/app/src/main/java/com/serwylo/lexica/db/migration/MigrateHighScoresFromPreferences.kt#L105) and [friends](https://github.com/lexica/lexica/blob/master/app/src/main/java/com/serwylo/lexica/db/migration/MigrateHighScoresFromPreferences.kt#L36)).

These two things could be pulled somewhere generic so that they could also be used for the generation of the QR code string.

The other good point from #213 is that sharing via SMS or email is also a good option, therefore the "Share game" should have the option of QR or NFC, but also just normal sharing using whatever app you wish via the normal sharing mechanism.

**Comment by ig9000** [comment]
I didn't manage to find this ticket before posting #213; additional thoughts from here:
- it would be neat if the export formatting can be human-readable to a reasonable extent, so that it could be shared with people who may not be using a platform which supports Lexica. In my imagination, it's something like this:
```q w e r t
   y u i o p
   a s d f g
   h j k l z
   x c v b n

L15 T5 S2
```
- Putting more work into the import parsing would allow all non-printing characters to be discarded when encoding to QR, etc, as well as providing more robustness against mangled formatting in general
- I have no idea if any similar apps offer anything like this already, but it might be worth looking into, since it would be interesting to have a "standard" which could even be exchanged between apps

Thanks again for reanimating this project!

**Comment by pserwylo** [comment]
Alright, this is probably the next big ticket item on my radar (it also facilitates #30 quite trivially once implemented). For those curious, here is a sketch of what I think it will end up looking like:

* [ ] Two play buttons on the home screen:
  * [ ] New game (same as currently, but with an icon of a single person to indicate single player).
  * [ ] Multiplayer (to the right of "New Game", with an icon of two or more people).
* [ ] Opening multiplayer will launch something similar to the F-Droid swap UI:
  * [ ] A prominent QR code will be shown (also supporting NFC).
  * [ ] Some text explaining "To join, scan this QR code (make sure you have Lexica installed). When all players are ready, start the game."
  * [ ] A button saying "Start game".
* [ ] When others scan, it opens Lexica and:
  * [ ] Shows the details of the game mode (e.g. duration, min word length, etc - similar to how it is shown in the game mode selection list).
  * [ ] Some text explaining "When all players are ready, start the game."
  * [ ] A button saying "Start game".

This will be implemented as a QR code that encodes not just the board, but also the language, and the relevant details of the game mode.

This needn't be in a human readable format for this mechanism of sharing:
```
lexica://share/ABCDEFGHIJKLMNOPQRSTUVWXY?l=fr_FR&t=2700&s=l&m=4&h=nc
```

Results in this nice compact (and thus easier to scan on low-end devices) QR code:

![qr-small](https://user-images.githubusercontent.com/248565/112749867-d86f9580-9010-11eb-971b-e698ff6718e9.png)

However, in my initial tests, I've also put together a human readable format:
```
lexica://share/
A B C D E
F G H I J
K L M N O
P Q R S T
U V W X Y

Language: fr_FR
Time: 45m
Score: Letter
Min Word Length: 4
Hints: Colour + Number
```

Which results in this QR code:

![qr-big](https://user-images.githubusercontent.com/248565/112749870-ddcce000-9010-11eb-8d1a-6e43f4b344a0.png)

**Comment by ig9000** [comment]
Nice. Such a human readable format would allow to share with somebody who may not have Lexica installed, but could still be imported if they do. Personally, I would also like to be able to save/export a game after the fact (in single-player mode), since it's often while playing one that something about it suggests that it would be fun for some other particular person.

**Comment by pserwylo** [comment]
> Such a human readable format would allow to share with somebody who may not have Lexica installed.

Makes sense, except there may be a few technical limitations we need to experiment with to understand if they are a blocker or not. We ran into this issue when implementing Swap for F-Droid too. It is difficult to make a single share URI that works for both those who have Lexica installed and those who don't. Some of the constraints here are:
 * URIs will generally need to encode whitespace to be valid, so `A B C D E` for the first row of a board will become `A%20B%20...` and each newline becomes `%0A`. Therefore, the "Human readable" version becomes a lot less human readable.
 * A `lexica://...` prefix in the URI is realistically needed to intercept such requests automatically (though not strictly required - it is a lot safer).

In the end, I think the simplest user experience for such a scenario is to have the main happy-path for people who have lexica installed (as described above), and then for those who don't have it installed, offer the following:

* [ ] A description + link saying: "Don't have Lexica installed?", which when clicked, opens up a popup with the following options:
  * [ ] A prominent link to "Download Lexica" which takes you to the playstore URI for Lexica (which F-Droid will happily intercept too and take you to the install page).
  * [ ] The text "Can't install Lexica? Scan this QR instead and play with a pen and paper."
  * [ ] A QR code with the human readable format.
  * [ ] A back button to return to the normal multiplayer screen (if all of this "Don't have lexica?" stuff is in a popup, then it should make it clear that closing the popup returns to the exact same multiplayer screen as before, which will retain the same random board which may already have been shared with others).

We could then also think about what it means in order to share via SMS or other means.

**Comment by frief** [comment]
Great, thanks!

it might be possible to reduce the number of keypresses/text shown without loosing much functionality:

* [ ]  One "Start game"  button on the home screen, QR code is always shown. In addition the number of words in the challenge is shown
* [ ]  When others scan, it opens Lexica and:
  * [ ]  Shows the very same screen as the primary player but with prominent notice that it's initialized remotely (eventually with a hash). No extra text to start simultaneously. One can probably rely on social interaction to enforce this anyway:) If not no harm is done as long as the player does not share results prematurely. And even if so the harm is limited to the very player that shares prematurely. Social interaction might be strong enough to remove the need for separate URI lexica://share/ and lexica://play/ (not sure). A moderate source of conflict is fine (maybe even intended - that's what games might also be about) in my view.
* [ ]  Pressing "Start game" will, unchanged from your proposal:
  * [ ]  Ensure an appropriate game mode exists in the database. If not, create a new custom one with a label of "Shared game".
  * [ ]  Setup a board with appropriate letters.
* [ ]  In the end of game scoring screen, show the QR code and a "Replay" button.

**Comment by ig9000** [comment]
> I think the simplest user experience for such a scenario is to have the main happy-path for people who have lexica installed

Makes sense not to try to overload the URL too much. I personally wouldn't mind going through a couple of menus to export in a less frequently used format, especially if it means being able to choose the desired amount of "promotional" verbosity, etc.

**Comment by pserwylo** [comment]
Thanks for the feedback all. Here is where I'm up to so far after hacking around a bit. Already thought through a few different alternatives just by trying them out, so looking forward to where further testing gets us:

![device-2021-04-06-125441](https://user-images.githubusercontent.com/248565/113652756-bd1f1d00-96d7-11eb-86d6-05961c3584b9.png) ![device-2021-04-06-125453](https://user-images.githubusercontent.com/248565/113652761-be504a00-96d7-11eb-81be-379f61f0a4e1.png) ![device-2021-04-06-125454](https://user-images.githubusercontent.com/248565/113652762-bee8e080-96d7-11eb-8835-958e177806b2.png) ![device-2021-04-06-125534](https://user-images.githubusercontent.com/248565/113652765-bf817700-96d7-11eb-832e-77bad423061a.png)

The "Share Invites" opens up a standard share dialog, which shares the message shown in the last screenshot.

Obviously all of this needs some polish, and there are a few quirks getting the game to start with a non-random board that I need to work through, but overall I think it will end up with quite a nice UX.

Some further thoughts:
* I'll replace the `lexica://` URI with a `https://lexica.github.io/lexica/multiplayer/` URI.
* This will allow for Lexica to intercept such URLs and automatically open the multiplayer lobby screen.
* If Lexica is not installed, or the user accidentally said "Open with browser" instead of "Open with lexica", then it will show two links:
  * Get lexica (F-Droid / Play Store)
  * Open in lexica (launches a `lexica://` URI, which will never open in a browser and will only work if you have lexica installed).
* This screen could also have some javasript which looks at the URL and renders the board to the screen behind a "Start game" button so that people can play with a pen and paper, but that is a bit out of scope here.

**Comment by pserwylo** [comment]
General question for any translators (I'll link to this post via a Weblate announcement too): I want to release a version of this to get some feedback, but I have no idea if the UI I have is going to be the final one.

Would you rather:
* Translate all strings, commit, have Weblate ask you to translate them. If the entire UI is overhauled soon after the release, then delete the strings and add new ones as required, to be translated again.
* Leave the strings for the new UI as untranslatable English for this early stage while seeking feedback. Once we receive feedback from the next release or two and the UI stabalises, **then** allow them to be translated via Weblate.

**Comment by comradekingu** [comment]
Ship early, ship often. "Translators" are also people picking out errors in the source strings, and in turn _want_ to see changes happen. It is a self-correcting system. "Translators" have no-one to blame but themselves.

The feeling of translating something that doesn't pass for mustard is nobody cares where it matters, and this is often reflected in subsequent translations. Instruct people to fix the source strings directly instead.

As an insider tip, shipping often enough gives the project more attention. Not only is it remembered from time to time, but it increases chances of translators getting subscribed, because that is now a result of changing just one string.

Shipping often factors out errors over a shorter timespan, which does bring more of the effort in line with something that sticks over time. If you at one point attract a rare language, make sure it can get finished in as much time there was to spare doing it.

Leaving strings untranslatable is the height of inefficiency, as it isn't much harder to translate and then give feedback because the insight is the precursor. This also grants the benefit of thinking about something in at least two languages before opining on it.

The people reacting the quickest are going to be the people doing the heavy lifting too. Consequently, you will be even more biased towards this by asking here. That is however OK, because facilitating it helps everyone.

TL;DR Smooth out problems sooner to lower the barrier to entry.

**Comment by pserwylo** [comment]
Thanks for the feedback @comradekingu, informative as always. Also makes perfect sense, I'm always after as much feedback as possible. I'll take that on board for this and future releases.

**Comment by frief** [comment]
Big thanks for implementing this! And it feels _really_ streamlined:)

Two minor requests
- the email notification (the html part of the mail) could use monospace font for the letter field.
- it would be nice if the number of words within the challenge would be known before sharing it. Depending on the number of words contained the character of the game changes for me (from being limited by letter connection speed to working hard to find the next word). This could be a number shown after the letter field is generated and before it is shared or before the letter field is generated (e.g. challenge with 100..149 words, ..))

**Comment by pserwylo** [comment]
Thanks for the feedback @frief.

Both comments seem very sensible. The first is straightforward (though keep in ind that this will never really be perfect - e.g. an English board with a "Q" is already stuffed up because it doesn't follow the rules, and to be honest it isn't at the top of my personal priority list to fix that for all possible languages supported.

The second one is a nice touch too. It will require a small additional tweak that not only shows the number of words, but also allows the board to be refreshed with new letters (without showing the words, just showing the number of words).

I'm going to start putting together a roadmap for Lexica and this will form one of the items.
